sys/include/xtimer.h

```diff
@@ -0,0 +1,442 @@
+/*
+ * Copyright (C) 2015 Kaspar Schleiser <kaspar@schleiser.de>
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @defgroup  sys_xtimer Timers
+ * @ingroup   sys
+ * @brief     Provides a high level timer module to register
+ *            timers, get current system time, and let a thread sleep for
+ *            a certain amount of time.
+ *
+ * The implementation takes one low-level timer that is supposed to run at 1MHz
+ * speed and multiplexes it.
+ *
+ * Insertion and removal of timers has O(n) complexity with (n) being the
+ * number of active timers.  The reason for this is that multiplexing is
```

Comment:
> You state this as if it was a new insight. :)

I stated this because you seem to have a different understanding.

> it would mean that currently never-fail functions might fail, forcing any user of xtimers to deal with a failing timer, making the overall codebase larger and more complex

Once again: I think this is a low price to pay. Checking a return value is not an incredible complex task. And what happens if there is no memory left for a new timers? *Never* failing is maybe not accurate.

> users of xtimer with realtime requirements will need to make sure that the system is not using the maximum amount of timers

Users also have to check that their application never exceeds the amount of available memory. Programming an embedded applications means that you have to know which state your system might have at any point. Hence,

>  a failed timer must be dealt with

is IMO not true, because a failed timer set indicates a critical system failure due to immature application design.